### PR TITLE
Use PETSc metric utils

### LINKED ISSUE #52: Use PETSc metric utils
I have been extending the metric-based mesh adaptation functionality of PETSc over the past few months (see [this tracker](https://gitlab.com/jwallwork23/petsc/-/wikis/PETSc-ParMmg-integration-tracker)). They are in the process of being piped through to Firedrake (see [this PR](https://github.com/firedrakeproject/firedrake/pull/2295)). It would be nice to use these and cut down the code in `pyroteus/metric.py`.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #52.

Much of the metric-based functionality can now be used on the PETSc level via Firedrake. As such, it is removed from Pyroteus and the tests are moved to Firedrake (on branch `jwallwork23/parmmg-metric-based`).

### CODE REVIEW

**Comment by jwallwork23** [comment]
@acse-sl420 @acse-ej321 @ddundo Adding you all as reviewers because I highly recommend that you look at the changes here and in the [most recent version of meshadapt.py](https://github.com/firedrakeproject/firedrake/blob/jwallwork23/parmmg-metric-based/firedrake/meshadapt.py) (for which I will be making a PR to Firedrake soon).

Don't worry about reviewing this in detail, I'm mainly posting here so that you are aware of the changes. Feel free to accept the changes if you are happy with them but please don't click 'merge' until everyone has done so. Also feel free to make any comments or ask any questions using the 'Files changed' tab :).

**Review by ddundo — APPROVED** [review]
Thanks @jwallwork23! Looks good :) Just a tiny unimportant comment on the docstring of `metric.density_and_quotients`: shouldn't the definition of density and quotients be

$$
        d = \prod_{i=1}^n \frac{1}{h_i},\qquad
        r_i = h_i^n d,\qquad \forall i=1:n?
$$

**Comment by jwallwork23** [comment]
> Thanks @jwallwork23! Looks good :) Just a tiny unimportant comment on the docstring of `metric.density_and_quotients`: shouldn't the definition of density and quotients be
> 
> d=∏i=1n1hi,ri=hind,∀i=1:n?

@ddundo Please could you post this again as a comment on the actual code? I can't find the lines you are referring to. As I said above, navigate to the 'Files changed' tab, selecting a snippet of code (by hovering over it with the mouse and using the blue + signs), typing a comment, and then clicking either "Add single comment" or "Start a review". Thanks!

File: pyroteus/metric.py [hunk 1]
```diff
@@ -623,126 +602,89 @@ def determine_metric_complexity(
...
 
-@PETSc.Log.EventDecorator("pyroteus.density_and_quotients")
-def density_and_quotients(metric: Function, reorder: bool = False) -> List[Function]:
+@PETSc.Log.EventDecorator()
+def density_and_quotients(
+    metric: RiemannianMetric, reorder: bool = False
+) -> List[Function]:
     r"""
     Extract the density and anisotropy quotients from a
     metric.
```
Comment (ddundo) [inline pyroteus/metric.py:690]: Definition of density and quotients

$$
        d = \prod_{i=1}^n \frac{1}{h_i},\qquad
        r_i = h_i^n d,\qquad \forall i=1:n?
$$

**Comment by ddundo** [comment]
@jwallwork23 Sorry, couldn't comment on the actual lines since they're unchanged (hence why I commented here :) ), but I commented on the line where you define the function to help find it.

File: pyroteus/metric.py [hunk 2]
```diff
@@ -420,29 +421,39 @@ def enforce_element_constraints(
     for metric, hmin, hmax, amax in zip(metrics, h_min, h_max, a_max):
         fs = metric.function_space()
         mesh = fs.mesh()
+        P1 = FunctionSpace(mesh, "CG", 1)
+
+        def interp(f):
```
Comment (acse-ej321) [inline pyroteus/metric.py:427]: @jwallwork23 Can you double check that this logic is the same as the previous code block? I am reading the original as if True, then clement_interpolant, whereas the new code seems to be opposite this interpretation.

File: pyroteus/metric.py [hunk 3]
```diff
@@ -420,29 +421,39 @@ def enforce_element_constraints(
...
+            integral = firedrake.assemble(ufl.conditional(hmax < hmin, 1, 0) * dx)
+            if not np.isclose(integral, 0.0):
+                raise ValueError(
+                    f"Encountered regions where hmax < hmin: volume {integral}."
+                )
+            _amax = amax.vector().gather().min()
+            if _amax < 1.0:
+                raise ValueError(f"Encountered amax value smaller than unity: {_amax}.")
 
         # Enforce constraints
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply pyroteus/metric.py:459]: Oh good spot! I'll add it back in.

File: pyroteus/metric.py (hunk 2 above)
Comment (jwallwork23) [reply pyroteus/metric.py:427]: Another good spot, thanks!

**Comment by jwallwork23** [comment]
@acse-ej321 Hopefully you find commit https://github.com/pyroteus/pyroteus/pull/110/commits/6ae680f4c8d013104c84529573098a4d9b684838 useful. It adds an `export` method to `MetricParameters` and `GoalOrientedMetricParameters` so that they can pass parameters to PETSc via the `RiemannianMetric`. (Note that I recently updated the Firedrake PR so that `set_parameters` updates them, rather than resetting before setting.)

**Comment by jwallwork23** [comment]
Also added a test. I think I've addressed all feedback so far but let me know if not @ddundo @acse-ej321

**Comment by ddundo** [comment]
Not sure what the best way is to ask you about changes on the Firedrake branch, but L241 in firedrake/meshadapt.py `return sol.subfunctions` should maybe be `return sol.split()`? I get `AttributeError: 'CoordinatelessFunction' object has no attribute 'subfunctions'` from the following:

```
Q = FunctionSpace(msh, "CG", 1)
u_mag = interpolate(sqrt(dot(u, u)), Q)

P1_ten = TensorFunctionSpace(msh, "CG", 1)
M = RiemannianMetric(P1_ten)
M.compute_hessian(u_mag)
```

where `u` is my velocity vector solution on mesh `msh` .

**Comment by jwallwork23** [comment]
> Not sure what the best way is to ask you about changes on the Firedrake branch, but L241 in firedrake/meshadapt.py `return sol.subfunctions` should maybe be `return sol.split()`? I get `AttributeError: 'CoordinatelessFunction' object has no attribute 'subfunctions'` from the following:
> 
> ```
> Q = FunctionSpace(msh, "CG", 1)
> u_mag = interpolate(sqrt(dot(u, u)), Q)
> 
> P1_ten = TensorFunctionSpace(msh, "CG", 1)
> M = RiemannianMetric(P1_ten)
> M.compute_hessian(u_mag)
> ```
> 
> where `u` is my velocity vector solution on mesh `msh` . Or am I not using these new changes right? :)

Are your Firedrake dependencies up to date? I believe UFL changed from `split` to `subfunctions`. Perhaps try `firedrake-update` and if that breaks, reinstall.

File: pyroteus/metric.py [hunk 4]
```diff
@@ -480,93 +476,73 @@ def enforce_element_constraints(
...
 
     # Normalise on each subinterval
     for metric, tau in zip(metrics, dt_per_mesh):
-        determinant = (
-            1 if p == "inf" else pow(tau**2 * ufl.det(metric), -1 / (2 * p + d))
-        )
-        metric.interpolate(global_norm * determinant * metric)
+        metric.set_parameters(metric_parameters)
+        metric.normalise(global_factor=pow(tau, -2 / (2 * p + d)) * global_factor)
+        metric.enforce_spd(**kwargs)
```
Comment (ddundo) [inline pyroteus/metric.py:545]: Would it make sense / be possible to use different metric parameters on different subintervals? In particular, this is motivated by my question last week on me wanting to ramp up complexity in simulation time

File: pyroteus/metric.py [hunk 5]
```diff
@@ -480,93 +477,84 @@ def enforce_element_constraints(
...
+        normalisation order under `dm_plex_metric_p`, or a list thereof
+    :kwarg global_factor: pre-computed global normalisation factor
+    :kwarg boundary: is the normalisation to be done over the boundary?
+    :kwarg restrict_sizes: should minimum and maximum metric magnitudes
+        be enforced?
+    :kwarg restrict_anisotropy: should maximum anisotropy be enforced?
+    """
+    if isinstance(metric_parameters, dict):
+        metric_parameters = [metric_parameters]
+    for mp in metric_parameters:
```
Comment (jwallwork23) [inline pyroteus/metric.py:507]: @ddundo Your comment seems to have vanished. Either way, I implemented what you requested. Hopefully looks good?

**Comment by jwallwork23** [comment]
@ddundo @acse-ej321 Please could you re-review this? Would be good to get it merged. Thanks.

File: pyroteus/metric.py (hunk 5 above)
Comment (ddundo) [reply pyroteus/metric.py:507]: Thanks @jwallwork23! And sorry about that - I was going to do it myself so I deleted to prevent you doing it :)

**Comment by ddundo** [comment]
Hey @jwallwork23, I changed `metric.space_time_normalise` to follow [Alauzet & Olivier (2010)](http://congress2.cimne.com/eccomas/proceedings/cfd2010/papers/01600.pdf) which you also describe in your thesis, but the implementation was quite different. I get results that I'd expect now, but please check (@acse-ej321 too please :) ). Do you know where you got your implementation?

File: pyroteus/metric.py [hunk 6]
```diff
@@ -503,7 +503,7 @@ def space_time_normalise(
     :kwarg restrict_anisotropy: should maximum anisotropy be enforced?
     """
     if isinstance(metric_parameters, dict):
-        metric_parameters = [metric_parameters]
+        metric_parameters = [metric_parameters for _ in range(len(time_partition))]
```
Comment (jwallwork23) [inline pyroteus/metric.py:506]: If you're not going to use the iterates then you could just do `[metric_parameters for _ in time_partition]`

File: pyroteus/metric.py [hunk 7]
```diff
@@ -546,14 +543,14 @@ def space_time_normalise(
             detM = ufl.det(metric)
             dX = (ufl.ds if boundary else ufl.dx)(metric.function_space().mesh())
             exponent = 0.5 if np.isinf(p) else (p / (2 * p + d))
-            integral += firedrake.assemble(pow(tau**2 * detM, exponent) * dX)
+            integral += firedrake.assemble(pow(detM, exponent) * dX) * pow(tau, -1)
```
Comment (jwallwork23) [inline pyroteus/metric.py:546]: Minor point: I'd rather have `/ tau` than `* pow(tau, -1)`.

**Comment by jwallwork23** [comment]
Okay I'm going to merge this and then we can follow up when addressing #36 and #115.

**Pull request merged**